src/python/pants/jvm/shading/jarjar.default.lockfile.txt

```diff
@@ -0,0 +1,26 @@
+# This lockfile was autogenerated by Pants. To regenerate, run:
+#
+#    ./pants generate-lockfiles
+#
+# --- BEGIN PANTS LOCKFILE METADATA: DO NOT EDIT OR REMOVE ---
+# {
+#   "version": 1,
+#   "generated_with_requirements": [
+#     "com.googlecode.jarjar:jarjar:1.3,url=not_provided,jar=not_provided"
+#   ]
+# }
+# --- END PANTS LOCKFILE METADATA ---
+
+[[entries]]
+directDependencies = []
+dependencies = []
+file_name = "com.googlecode.jarjar_jarjar_1.3.jar"
+
+[entries.coord]
+group = "com.googlecode.jarjar"
+artifact = "jarjar"
+version = "1.3"
+packaging = "jar"
```

Comment:
Would suggest using https://github.com/eed3si9n/jarjar-abrams instead: pantsbuild used to have a fork of jarjar, and they reached out a while back to ask whether we'd like to collaborate on their fork instead.

Reply:
surprisingly, when moved to the `jarjar-abrams` implementation I'm getting duplicate JAR entry exceptions from the `jarjar` process for the `META-INF/MANIFEST.MF` and `META-INF/LICENSE` files.

The former is easy to workaround, as we can tell `jarjar` to ignore the manifest file, the other one is not so simple and may require some duplicate resolution strategy in a similar way as they do in the `sbt-assembly` plugin.

Reply:
Ok: I'm fine with punting, but if you're able to open an issue with them, that would be good.

IIRC, `jarjar` from google code does not support more modern JDKs, so it will be necessary to get on a modern version eventually.

Reply:
did a bit of digging and, while the fat jars produced by our `deploy_jar` target are runnable using `java -jar fatjar.jar`, they do indeed have duplicate entries inside of them. This can be verified in a number of ways, the simplest of them is trying to unzip one of them and observing how `unzip` keeps asking us to take a decision on what to do with each of the duplicated entries.

I will revert this back to the google jarjar version as a way of getting the feature implemented. But will add a ticket on how to investigate possible solutions to do deduplication when merging the uberjars (and upgrade the `jarjar` version afterwards).

Reply:
Ah, yea: that is expected given how we implement the JAR merge (which is nearly just a "concat" of the JAR files, followed by an index fixup). That operation is quite efficient, so ideally we don't have to actually handle dupes there... dupes are legal inside of JARs as long as the index is valid. So possibly worth a bug report with https://github.com/eed3si9n/jarjar-abrams ...?

Reply:
hitting a bit of a roadblock now, tried three different forks (including the one suggested here before) but all of them fail to shade the JAR if they find duplicate entries. Only the `jarjar-abrams` shows the failure at the process level, the rest of them silently fail showing the error in stderr but with a success exit code.

Think this will need to be parked unless I fork one of the forks and remove that condition for not having duplicates...

Will try to look for the Pants v1 code that did the same to see if I can get some ideas.

Reply:
> Will try to look for the Pants v1 code that did the same to see if I can get some ideas.

The Pants v1 code used a different strategy to achieve the same "concatenate a jar file without recompressing it" effect... it may not have resulted in duplicate entries the way that this one does.

It would be an option to change zipping strategies when shading will be used, to first extract all smaller jars and then re-zip them (although it would mean re-compressing).

Reply:
> > Will try to look for the Pants v1 code that did the same to see if I can get some ideas.
> 
> The Pants v1 code used a different strategy to achieve the same "concatenate a jar file without recompressing it" effect... it may not have resulted in duplicate entries the way that this one does.

Concretely: it used some Java code to concatenate jar files by copying the content of entries without recompressing it: it still exists on the `1.30.x` branch, but we went with the `zip -FF` approach for expediency: https://github.com/pantsbuild/pants/blob/1.30.x/src/java/org/pantsbuild/tools/jar/Main.java

Reply:
cool, I guess that rescuing that code can help with this issue without impacting too much the efficiency of the far jar creation.

I will look at this over the next days and see what I can do about it.